File: JAVA/1249. Minimum Remove to Make Valid Parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {        
        char [] chars = new char[s.length()];
        for(int i = 0; i < s.length(); i++)chars[i] = s.charAt(i);
        
        Stack<Integer> st = new Stack<>();
        for(int i = 0; i < s.length(); i++){
            if(chars[i] == '(')st.push(i);
            else if(chars[i] == ')'){
                if(st.size() == 0)chars[i] = '.';
                else st.pop();
            }
        }
        while(st.size() != 0)chars[st.pop()] = '.';
        
        StringBuilder ans = new StringBuilder();
        for(int i = 0; i < chars.length; i++){
            if(chars[i] != '.')ans.append(chars[i]);
        }
        return ans.toString();
            
    }
}

File: C++/1249. Minimum Remove to Make Valid Parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        int close = 0, diff = 0; 
        for (auto& ch : s) 
            if (ch == ')') ++close;        
        string ans; 
        for (auto& ch : s) {
            if (ch == '(') {
                if (diff < close) ++diff; 
                else continue; 
            } else if (ch == ')') {
                --close; 
                if (diff > 0) --diff; 
                else continue; 
            }
            ans.push_back(ch); 
        }
        return ans; 
    }
};
